slashing/program/src/state.rs

```diff
@@ -0,0 +1,121 @@
+//! Program state
+use {
+    crate::error::SlashingError,
+    bytemuck::{Pod, Zeroable},
+    solana_program::{program_pack::IsInitialized, pubkey::Pubkey},
+};
+
+const PACKET_DATA_SIZE: usize = 1232;
+
+/// Types of slashing proofs
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum ProofType {
+    /// Proof consisting of 2 shreds signed by the leader indicating the leader
+    /// submitted a duplicate block.
+    DuplicateBlockProof,
+    /// Invalid proof type
+    InvalidType,
+}
+
+impl ProofType {
+    /// Size of the proof account to create in order to hold the proof data
+    /// header and contents
+    pub fn proof_account_length(&self) -> Result<usize, SlashingError> {
+        match self {
+            // Duplicate block proof consists of 2 shreds
+            Self::DuplicateBlockProof => Ok(2 * PACKET_DATA_SIZE + ProofData::WRITABLE_START_INDEX),
+            Self::InvalidType => Err(SlashingError::InvalidProofType),
+        }
+    }
+}
+
+impl From<ProofType> for u8 {
+    fn from(value: ProofType) -> u8 {
+        match value {
+            ProofType::DuplicateBlockProof => 0,
+            ProofType::InvalidType => u8::MAX,
+        }
+    }
+}
+
+impl From<u8> for ProofType {
+    fn from(value: u8) -> Self {
+        match value {
+            0 => Self::DuplicateBlockProof,
+            _ => Self::InvalidType,
+        }
+    }
+}
+
+/// Header type for proof data
+#[repr(C)]
+#[derive(Clone, Copy, Debug, PartialEq, Pod, Zeroable)]
+pub struct ProofData {
+    /// Struct version, allows for upgrades to the program
+    pub version: u8,
```

Comment:
We probably don't need versioning for this -- we would likely create new proof types and deprecate the old ones as needed